### ISSUE TITLE
Envoy as a dynamic forward proxy with Host and Port rewrite

### ISSUE THREAD

**Issue by mabukhovsky** [envoyproxy/envoy#8768]
We would like to use Envoy as a forward proxy that handles connection to 3rd party services via HTTP / HTTPs requests that could point to different domains, ports, for example:
```
www.youtube.com:443/feed/trending
```
In our initial integration we were able to achieve the behavior we want by creating route, routeAction and cluster for every URL
```
"dynamic_active_clusters": [
        {
          "version_info": "2",
          "cluster": {
            "name": "fcaa46cf-a30b-4454-b0ed-170e97ed9c3c_cluster",
            "type": "STRICT_DNS",
            "connect_timeout": "5s",
            "tls_context": {
              "sni": "www.yotube.com"
            },
            "dns_refresh_rate": "3600s",
            "dns_lookup_family": "V4_ONLY",
            "load_assignment": {
              "cluster_name": "fcaa46cf-a30b-4454-b0ed-170e97ed9c3c_cluster",
              "endpoints": [
                {
                  "lb_endpoints": [
                    {
                      "endpoint": {
                        "address": {
                          "socket_address": {
                            "address": "www.youtube.com",
                            "port_value": 443
                          }
                        }
                      }
                    }
                  ]
                }
              ]
            }
          },
          "last_updated": "2019-10-25T17:01:42.507Z"
        }
      ]
    },

"dynamic_route_configs": [
        {
          "version_info": "2",
          "route_config": {
            "name": "primary_route_configuration",
            "virtual_hosts": [
              {
                "name": "someVirtualHost",
                "domains": [
                  "*"
                ],
                "routes": [
                  {
                    "match": {
                      "prefix": "/prefix/",
                      "headers": [
                        {
                          "name": "x-some-host",
                          "exact_match": "www.youtube.com"
                        },
                        {
                          "name": "x-some-proto",
                          "exact_match": "https"
                        },
                        {
                          "name": "x-some-port",
                          "exact_match": "443"
                        }
                      ]
                    },
                    "route": {
                      "cluster": "fcaa46cf-a30b-4454-b0ed-170e97ed9c3c_cluster",
                      "prefix_rewrite": "/",
                      "host_rewrite": "www.youtube.com"
                    }
                  },
```
While this approach works, there's a performance impact as Envoy has to find matching route, and if number of routes is significantly high (100K) Envoy have to go over loop of all routes if the valid route was added last in Envoy config.

Later we got requirement to support REST URLs which contain arguments, like:
```
www.google.com:443/maps/<location>/language/<language>
```
Dynamic forward proxy could solve the problem, as it allows forwarding to any domain + we can have auto_host_rewrite_header option in a route so for a call <envoy-host>://<some-path> forward will be done to the host specified in a header.

```
"dynamic_route_configs": [
        {
          "version_info": "2",
          "route_config": {
            "name": "primary_route_configuration",
            "virtual_hosts": [
              {
                "name": "someVirtualHost",
                "domains": [
                  "*"
                ],
                "routes": [
                  {
                    "match": {
                      "prefix": "/"
                    },
                    "route": {
                      "cluster": "dynamic_forward_proxy_cluster",
                      "auto_host_rewrite_header": "x-some-host"
                    }
                  }
                ]
              }
            ],
```
This approach is much better as we don't have define cluster and routes for all URL's like we did initially. It would be sufficient to have one  But with this approach, we don't see an option to provide PORT rewrite, which means we will still have to create lots of dynamic forward clusters, one cluster per (domain + port) combination. Ideally we wanted it to be an extra option on route, like "auto_port_rewrite_header".
Please reply and let us know if there's a way to achieve the same behavior differently (dynamic forward proxy with port and host rewrite).

**Comment by mattklein123** [comment]
cc @derekargueta @rgs1 who might be able to advise.

**Comment by rgs1** [comment]
@mabukhovsky so note that `host_rewrite` at the route action level won't do what you want, you'd need to use `host_rewrite`  via the `per_filter_config` for the dyn forward proxy:

https://github.com/envoyproxy/envoy/pull/8619

Extending the above to use a header value as the host rewrite target should be straight forward.

Port rewrites should work too, I think.

When you say http _and_ https, you mean for outgoing traffic right? For rewrites to work you can only do http:// otherwise for https:// it would be HTTP tunneling at which point Envoy can't rewrite the host header...

**Comment by rgs1** [comment]
I am happy to extend #8619 to support header value -> host rewrite, assuming that would solve your use case.

**Comment by rgs1** [comment]
I think setting `x-set-host: foo:81`'s value as `Host:` should handle host/port... No need for a separate header.

**Comment by mabukhovsky** [comment]
Thank you for clarification and quick responses guys, I truly appreciate it. 
```
per_filter_config:
                    envoy.filters.http.dynamic_forward_proxy:
                      host_rewrite: www.example.org
```
Yes, I've seen this option but as I've mentioned we want host and port values to be dynamic and taken from request header. Something like
```
per_filter_config:
                    envoy.filters.http.dynamic_forward_proxy:
                      host_rewrite_header: x-set-host
                      port_rewrite_header: x-set-port
```
or if you allow both host and port to be part of the rewrite header (header content example: google.com:443) as you've mentioned - that's awesome too.
Please let me know if that's something you could introduce soon, this feature is highly appreciated! We will switch to master immediately - that's how we need it.

**Comment by rgs1** [comment]
We should probably call it `auto_host_rewrite_header`, for consistency: 

https://github.com/envoyproxy/envoy/blob/master/api/envoy/api/v2/route/route.proto#L760

**Comment by mabukhovsky** [comment]
Let me give you little more detail about our goal. We don't have hosts in our configuration because we want Envoy to behave like generic forward proxy for HTTP/HTTPS requests with no restrictions on host/port etc at Envoy configuration as we will have millions of URLs to forward. All we want - one wildcard/regex virtual host / route and dynamic forward proxy cluster in configuration. Request should be forwarded with domain/port and protocol rewritten based on the data in request header (x-host). Even if our feature request is honored, it looks like we won't be able to forward HTTPS requests due to tunneling constraint. Is there any other way we can achieve this goal with Envoy without necessity to create cluster \ route for every URL we want to allow? 

**HTTP example**
```
http://<envoy-ip>:<envoy-port>/some-path with 
'auto_host_rewrite_header' set to "www.google.com:100" 
something like 'auto_protocol_rewrite_header' set to "http" 
should be forwarded by Envoy to http://www.google.com:100/some-path
```
**HTTPS example**
```
http://<envoy-ip>:<envoy-port>/some-path with 
'auto_host_rewrite_header' set to "www.google.com:443" 
something like 'auto_protocol_rewrite_header' set to "https" 
should be forwarded by Envoy to https://www.google.com:443/some-path
```

**Comment by rgs1** [comment]
@mabukhovsky note that the dynamic forward proxy will take a request with http:// and fetch it with https://, if the cluster has a `tls_context` set, e.g.:

https://gist.github.com/rgs1/772096cf9890108575909f9ba5ee76cf

You can test the above with something like:

```
export http_proxy=localhost:10000
curl -svo /dev/null http://www.google.com
```

which per the example in my config will get the host rewritten to www.pinterest.com...

So assuming this covers your use case correctly (that is, your incoming requests are http:// and you want them to be fetched as https:// -- so you'll configure a `tls_context`), I can send a follow-up PR for #8619 to support `auto_host_rewrite_header` in the dynamic forward proxy per filter config.

Does that sound good?

**Comment by mabukhovsky** [comment]
Thank you for replying, yes your configuration is very close to what we need.
What is a time frame for adding **auto_host_rewrite_header** filter feature?
Is it something your team can deliver in November, or it is a long term feature?

**Comment by rgs1** [comment]
@mabukhovsky @mattklein123 something like #8869?

**Comment by rgs1** [comment]
Ah hold on, port overwrite won't quite work yet, a small change is needed for that... will push another commit.

**Comment by rgs1** [comment]
Ok, we are good to go -- the dns cache will handle extracting the port from the host value.

**Comment by mabukhovsky** [comment]
Is this feature included into 1.12.0 release? Can you point me to fixing pull request, please?
